buffer/src/main/java/io/netty/buffer/api/BufferAccessors.java

```diff
@@ -0,0 +1,614 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api;
+
+/**
+ * This interface is just the primitive data accessor methods that {@link Buffer} exposes.
+ * It can be useful if you only need the data access methods, and perhaps wish to decorate or modify their behaviour.
+ * Usually, you'd use the {@link Buffer} interface directly, since this lets you properly control the buffer reference
+ * count.
+ */
+public interface BufferAccessors {
```

Comment:
I see no little endian access methods. Is it intended? How would a user read/write little-endian values?

Reply:
Yes, that is intended. To perform little-endian accesses, you change the byte order of the buffer with the `order(ByteOrder)` method. This method is a simple field-setter, so it's very fast.

Reply:
I found having `order(ByteOrder)` method starts to get very confusing to a binary codec maintainer when his or her codec is written to assume all buffers it handles have certain byte order and that was why we introduced a bunch of little endian accessors in favor of switching endians of buffers. Unless there's a strong use case of making endianness switchable, I'd like to ask for removing `order()` and `order(ByteOrder)` method so we don't make the same mistakes I made in the past with Netty 4 buffer API.

Reply:
This would make the API quite wide. The number of accessor methods would double, or we'd have to pass a byte order argument with every call. There used to be LE and BE accessor method variants https://github.com/netty/netty-incubator-buffer-api/commit/b5ac0fdbe6f99e82096486e87659116328db08c3 but I removed them to make the API less cluttered.

Reply:
How about adding a method like `littleEndian()` which returns an object with little endian accessors?

e.g.

```java
buffer.littleEndian().getShortLE(...)
```

This way, most users will not be overloaded with wide API surface while keeping as safe as my original suggestion.

Reply:
@trustin what would be the difference to what we have today with `order(ByteOrder...)` ?

Reply:
Alternatively… do we need to support little-endian access at all? There are very few usages of little-endian accessors in Netty itself today, and most of them are in the compression codecs, processing meta-data. Buffers could be exclusively BE and then use sites reverse the bytes as necessary.

Reply:
> @trustin what would be the difference to what we have today with order(ByteOrder...) ?

It prevents us from having both `getInt()` and `getIntLE()` in our internal implementation for example. This means a user that uses little-endian may have worse performance due to unnecessary allocation of little endian view object.

> Alternatively… do we need to support little-endian access at all? There are very few usages of little-endian accessors in Netty itself today, and most of them are in the compression codecs, processing meta-data. Buffers could be exclusively BE and then use sites reverse the bytes as necessary.

Wouldn't it cause poor performance when implementing a little endian protocol on a little endian CPU, because of double swap?

I'm fine with starting without having little endian support, i.e. no `order()`, but I think we should design the API carefully when we add support for it.

Reply:
> Wouldn't it cause poor performance when implementing a little endian protocol on a little endian CPU, because of double swap?

PC and server class CPUs (x86_64, Armv6+, Power) have instructions to either load/store with specific endian, or to flip endian of words. I expect it'll have limited impact on our pipelined OOO CPUs.

Reply:
Personal preference, but `Accessors` makes it sound like a utility class.   Consider making it singular `Accessor`.    (e.g. Buffer _is a_ BufferAccessor).